app/services/api/teachers/query.rb

```diff
@@ -0,0 +1,156 @@
+module API::Teachers
+  class Query
+    include Queries::FilterIgnorable
+
+    attr_reader :scope, :lead_provider_id
+
+    def initialize(
+      lead_provider_id: :ignore,
+      contract_period_years: :ignore,
+      training_status: :ignore,
+      api_from_teacher_id: :ignore,
+      updated_since: :ignore,
+      sort: { created_at: :asc }
+    )
+      @lead_provider_id = lead_provider_id
+      @scope = Teacher.distinct
+
+      where_lead_provider_is(lead_provider_id)
+      where_contract_period_year_in(contract_period_years)
+      where_training_status_is(training_status)
+      where_api_from_teacher_id_is(api_from_teacher_id)
+      where_updated_since(updated_since)
+      set_sort_by(sort)
+    end
+
+    def teachers
+      preload_associations(block_given? ? yield(scope) : scope)
+    end
+
+    def teacher_by_api_id(api_id)
+      return preload_associations(scope).find_by!(api_id:) if api_id.present?
+
+      fail(ArgumentError, "api_id needed")
+    end
+
+    def teacher_by_id(id)
+      return preload_associations(scope).find(id) if id.present?
+
+      fail(ArgumentError, "id needed")
+    end
+
+  private
+
+    def preload_associations(results)
+      preloaded_results = results
+        .strict_loading
+        .eager_load(
+          :teacher_id_changes,
+          lead_provider_metadata: {
```

Comment:
can we add `earliest_ect_at_school_period` and `earliest_mentor_at_school_period`